scapy/contrib/isotp.py

```diff
@@ -1611,3 +1614,375 @@ def recv(self, x=0xffff):
 
 if USE_CAN_ISOTP_KERNEL_MODULE:
     ISOTPSocket = ISOTPNativeSocket
+
+
+# ###################################################################
+# #################### ISOTPSCAN ####################################
+# ###################################################################
+def send_multiple_ext(sock, ext_id, packet, number_of_packets):
+    """ Send multiple packets with extended addresses at once
+
+    Args:
+        sock: socket for can interface
+        ext_id: extended id. First id to send.
+        packet: packet to send
+        number_of_packets: number of packets send
+
+    This function is used for scanning with extended addresses.
+    It sends multiple packets at once. The number of packets
+    is defined in the number_of_packets variable.
+    It only iterates the extended ID, NOT the actual ID of the packet.
+    This method is used in extended scan function.
+    """
+    end_id = min(ext_id + number_of_packets, 255)
+    for i in range(ext_id, end_id + 1):
+        packet.extended_address = i
+        sock.send(packet)
+
+
+def get_isotp_packet(identifier=0x0, extended=False):
+    """ Craft ISO TP packet
+    Args:
+        identifier: identifier of crafted packet
+        extended: boolean if packet uses extended address
+    """
+
+    if extended:
+        pkt = ISOTPHeaderEA() / ISOTP_FF()
+        pkt.extended_address = 0
+        pkt.data = b'\x00\x00\x00\x00\x00'
+    else:
+        pkt = ISOTPHeader() / ISOTP_FF()
+        pkt.data = b'\x00\x00\x00\x00\x00\x00'
+
+    pkt.identifier = identifier
+    pkt.message_size = 100
+    return pkt
+
+
+def filter_periodic_packets(packet_dict, verbose=False):
+    """ Filter for periodic packets
+
+    Args:
+        packet_dict: Dictionary with Send-to-ID as key and a tuple
+        (received packet, Recv_ID)
+        verbose: Displays further information
+
+    ISOTP-Filter for periodic packets (same ID, always same timegap)
+    Deletes periodic packets in packet_dict
+    """
+
+    delete_set = set()
+    filter_dict = {}
+
+    for value in packet_dict.values():
+        pkt = value[0]
+        idn = value[1]
+        if idn not in filter_dict:
+            filter_dict[idn] = [pkt]
+        else:
+            filter_dict[idn] += [pkt]
+
+    for idn in filter_dict:
+        pkt_lst = filter_dict[idn]
+        if len(pkt_lst) < 3:
+            continue
+
+        tg = [p1.time - p2.time for p1, p2 in zip(pkt_lst[1:], pkt_lst[:-1])]
+        if all(abs(t1 - t2) < 0.001 for t1, t2 in zip(tg[1:], tg[:-1])):
+            if verbose:
+                print("[i] Identifier 0x%03x seems to be periodic. "
+                      "Filtered.")
+            delete_set.add(idn)
+
+    for key in packet_dict.copy():
+        if packet_dict[key][1] in delete_set:
+            del packet_dict[key]
```

Comment:
What would be cleaner would be to add the key of `value` as a second value (tuple?) in `filter_dict`, which would allow you to iterate over those instead to delete the values of `packet_dict.

This would avoid the above copy/iteration. If you don't see what I mean don't bother, it's picky.

Reply:
I did a rework

Reply:
maybe such a filter mechanism would be interesting for `plist`. I'll check, if I can switch to a `plist`. Than I could add that filter to `plist` if you think it might be useful

Reply:
I'm not sure if you mean a general filter, or this specific one (which is too specific to be in `plist`).
I don't think it's needed: there already is a `filter()` function, and it's usually a better practice to use `filter` within `sniff()` itself

Reply:
I agree